types/bluebird/index.d.ts

```diff
@@ -44,10 +45,8 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
   /**
    * Promises/A+ `.then()`. Returns a new promise chained from this promise. The new promise will be rejected or resolved dedefer on the passed `fulfilledHandler`, `rejectedHandler` and the state of this promise.
    */
-  then<U1, U2>(onFulfill: (value: R) => U1 | Bluebird.Thenable<U1>, onReject: (error: any) => U2 | Bluebird.Thenable<U2>): Bluebird<U1 | U2>;
-  then<U>(onFulfill: (value: R) => U | Bluebird.Thenable<U>, onReject: (error: any) => U | Bluebird.Thenable<U>): Bluebird<U>;
-  then<U>(onFulfill: (value: R) => U | Bluebird.Thenable<U>): Bluebird<U>;
-  then(): Bluebird<R>;
+  // Based on PromiseLike.then, but returns a Bluebird instance.
+  then<TResult1 = R, TResult2 = never>(onfulfilled?: ((value: R) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): Bluebird<TResult1 | TResult2>;
```

Comment:
@sandersn Thanks for filing the bug. 🙂 This should make special-case overloads like this one (`then<U>`) mostly unnecessary, right?

(cc @andy-ms) In case of Bluebird it might still be beneficial to keep it anyways though, since it's practically quite easily possible to create long chain of (simplified):
```ts
promise
    .then(val => val, () => alternative)
    .then(val => val, () => alternative)
    .then(val => val, () => alternative)
    ...
```

which would then result in such a "giant union" type (`TVal | TAlternative | TAlternative | ...`).